alf/nest/cnest/nest.cpp

```diff
@@ -0,0 +1,494 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pybind11/pybind11.h>
+
+#include <algorithm>
+#include <exception>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace py = pybind11;
+
+template <typename... Args>
+std::string py_format(const std::string& format, Args... args) {
+  // Use python's string format
+  static py::object py_format = py::module::import("format").attr("format");
+  return py::cast<py::str>(py_format(format, args...));
+}
+
+bool IsNested(const py::object& value) {
+  return (py::isinstance<py::list>(value) || py::isinstance<py::tuple>(value) ||
+          py::isinstance<py::dict>(value));
+}
+
+bool IsNamedtuple(const py::object& value) {
+  return (py::isinstance<py::tuple>(value) && py::hasattr(value, "_fields"));
+}
+
+bool IsUnnamedtuple(const py::object& value) {
+  return (py::isinstance<py::tuple>(value) && !py::hasattr(value, "_fields"));
+}
+
+void AssertSameType(const py::object& value1, const py::object& value2) {
+  bool type_equal = value1.get_type().equal(value2.get_type());
+  bool both_dict =
+      (py::isinstance<py::dict>(value1) && py::isinstance<py::dict>(value2));
+  if (!(type_equal || both_dict)) {
+    throw std::runtime_error(
+        py_format("Different types! {} <-> {}", value1, value2));
+  }
+}
+
+void AssertSameLength(const py::object& seq1, const py::object& seq2) {
+  bool both_iterable = (py::isinstance<py::iterable>(seq1) &&
+                        py::isinstance<py::iterable>(seq2));
+  if (!both_iterable) {
+    throw std::runtime_error(py_format(
+        "The arguments should be iterable! seq1: {}, seq2: {}", seq1, seq2));
+  }
+  if (py::len(seq1) != py::len(seq2)) {
+    throw std::runtime_error(
+        py_format("Different lengths! {} <-> {}", seq1, seq2));
+  }
+}
+
+py::list ExtractFieldsFromNest(const py::object& nest) {
+  bool is_dict = py::isinstance<py::dict>(nest);
+  if (!(is_dict || IsNamedtuple(nest))) {
+    throw std::runtime_error(
+        py_format("Nest {} must be a dict or namedtuple!", nest));
+  }
+
+  std::vector<std::string> fields;
+  py::dict nest_dict;
+  if (is_dict) {
+    nest_dict = py::cast<py::dict>(nest);
+    for (const auto& item : nest_dict) {
+      fields.push_back(py::str(item.first));
+    }
+  } else {
+    for (const auto& field : nest.attr("_fields")) {
+      fields.push_back(py::str(field));
+    }
+  }
+  std::sort(fields.begin(), fields.end());
+
+  py::list ret;
+  for (const auto& field : fields) {
+    if (is_dict) {
+      auto key = py::str(field);
+      if (!nest_dict.contains(key)) {
+        throw std::runtime_error(py_format(
+            "Only support string keys in a dictionary!! Wrong key: {}", key));
+      }
+      ret.append(py::make_tuple(key, nest_dict[key]));
```

Comment:
If I use c++ structure, I think in the end when pybind converts this to python code, it still will convert c++ pair to python tuple?

Reply:
Because _MapStructure can access the  C++ structure more efficiently

Reply:
I tried using vector and c++ make_pair. This basically makes the whole typing much more difficult. Because the inputs have uncertain types of the values, I need to make template for the returned value and it seems a disaster in the end where the code cannot deduce the type automatically.